contracts/src/Teleporter/upgrades/TeleporterRegistry.sol

```diff
@@ -5,28 +5,109 @@
+ */
+contract TeleporterRegistry {
+    // Address that the off-chain Warp message sets as the "source" address.
+    // The address is not owned by any EOA or smart contract account, so it
+    // cannot possibly be the source address of any other Warp message emitted by the VM.
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    WarpMessenger public constant WARP_MESSENGER =
+        WarpMessenger(0x0200000000000000000000000000000000000005);
+
+    bytes32 public immutable blockchainID;
+
+    // The latest protocol version. 0 means no protocol version has been added, and isn't a valid version.
+    uint256 public latestVersion;
+
+    // Mappings that keep track of the protocol version and corresponding contract address.
+    mapping(uint256 => address) internal _versionToAddress;
+    mapping(address => uint256) internal _addressToVersion;
 
     /**
-     * @dev Gets the {ITeleporterMessenger} contract of the given `version`.
+     * @dev Emitted when a new protocol version is added to the registry.
+     */
+    event AddProtocolVersion(
+        uint256 indexed version,
+        address indexed protocolAddress
+    );
+
+    /**
+     * @dev Initializes the contract by setting `blockchainID` and `latestVersion`.
+     * Also adds the initial protocol versions to the registry.
+     */
+    constructor(ProtocolRegistryEntry[] memory initialEntries) {
+        blockchainID = WARP_MESSENGER.getBlockchainID();
+        latestVersion = 0;
+
+        for (uint256 i = 0; i < initialEntries.length; i++) {
+            _addToRegistry(initialEntries[i]);
+        }
+    }
+
+    /**
+     * @dev Gets and verifies a Warp off-chain message, and adds the new protocol version
+     * address to the registry.
+     * If a version is greater than the current latest version, it will be set as the latest version.
+     * If a version is less than the current latest version, it is added to the registry, but
+     * doesn't change the latest version.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
      * Requirements:
      *
-     * - `version` must be a valid version, i.e. greater than 0 and not greater than the latest version.
+     * - a valid Warp off-chain message must be provided.
+     * - source chain ID must be the same as the blockchain ID of the registry.
+     * - origin sender address must be the same as the `VALIDATORS_SOURCE_ADDRESS`.
+     * - destination chain ID must be the same as the blockchain ID of the registry.
+     * - destination address must be the same as the address of the registry.
+     * - version must not be zero.
+     * - version must not already be registered.
+     * - protocol address must not be zero address.
      */
-    function getTeleporterFromVersion(
-        uint256 version
-    ) external view returns (ITeleporterMessenger) {
-        return ITeleporterMessenger(_getAddressFromVersion(version));
+    function addProtocolVersion(uint32 messageIndex) external {
```

Comment:
When you update this branch with `main`, `message.destinationAddress` will no longer be available.

I think we need to have some part of the off-chain message contain `address(this)` still though, in order to support multiple registry contracts on a single chain that should be updated independently. My suggestion would be for the message payload to be something like `abi.encode(registryAddress, ProtocolRegistryEntry);` so then we can decode and check the registryAddress the message is intended for.

An alternative would be to require that `message.originSenderAddress == address(this)`, since it isn't possible for this contract to send messages either. That would remove the need for `VALIDATORS_SOURCE_ADDRESS`, but may be slightly less clear convention.

Reply:
same thoughts on the first approach. Think the latter is more confusing in terms of "why is this contract sending a message to itself", etc.